### ISSUE TITLE
Add System.Security.Cryptography.Xml.SignedXml class

### ISSUE THREAD

**Issue by ghost** [dotnet/corefx#4278]
## Execution plan

**Goal:** Provide APIs fully compatible with full/Desktop .NET Framework (no changes as part of this work - straight port only)

**Plan:**
- [x] 1. Add the source code on GH sanitized, with licenses, etc. (it won't build)
- [x] 2. Make the source code build (still excluded from overall repo build)
- [x] 3. Remove Desktop registry-compat code paths
    - Remove methods that have [RegistryPermission] (Utils and SignedXml classes) along with any owned methods
    - Address any other registry-related compile errors such as those using Registry, RegistryPermission, RegistryKey, etc (delete code as necessary)
- [ ] 4. Add tests (We have to agree on expected code coverage and how much of the spec we have to cover)
    - Compare test results between Desktop and .NET Core implementations
- [x] 5. Make it part of overall repo build & ship it!

Code changes rules: **Only code changes absolutely necessary** to make it build and compatible with (full) .NET Framework are allowed, no additional bug fixes or changing architecture - such PRs will be rejected now.
Changes like that can be considered after the initial port is done, when we have a good test bed and when we are able to validate such change.

If larger code/architecture changes are needed from some reason, we should discuss them first here, before doing the work / submitting PR.

If you work on some parts of the plan, please say so in the discussion to avoid duplicated work. We (@steveharter @karelz) will co-assign the issue to you.

---------------------------------------------------------
## Original

The class to digitally sign XML needs to be added.

**Comment by rschiefer** [comment]
As referenced by Tratcher, this is a blocker for adding support for WsFederation/ADFS in ASP.NET 5.  We use ADFS extensively for many enterprise ASP.NET 4 applications.  We are very interested in migrating to ASP.NET 5 and using WsFederation.

**Comment by bartonjs** [comment]
@rschiefer @Tratcher Well, it's... complicated.

* ADFS doesn't use System.Security.Cryptography.Xml.SignedXml; but instead its own implementation.
    * (That's largely from dusting off mental cobwebs and remembering back to being on that team for version 1 of ADFS)
* System.IdentityModel definitely doesn't use System.Security.Cryptography.Xml
    * (That's 'cuz their source code on referencesource says so :smile:)
* People don't really like System.Security.Cryptography.Xml.SignedXml because it's based on XmlDocument, which leads to some perf issues.
    * The ADFS version was based on XmlReader, IIRC.
* So, likely what people want is CoreFX to create a new implementation of SignedXml.
* But a new version of SignedXml won't compatible with the old version of SignedXml
* So some people might want us to keep the old version, too.
* So in all, people really want two versions.
* Except they don't want the old version.
* Except for when they do.

So we're left with the conundrum of:
* We can port that class that no one really wants
* Or, we can stop and design a brand new thing that probably no one will use, since they can't be compatible with anything else
* Or, to use the most effort and still not benefit anyone, do both :smile:.

If the API surface of ADFS/WCF/etc permitted them to have a private implementation in the past, they should be able to continue doing so while moving to .NET Core.

@terrajobst - fyi

**Comment by bartonjs** [comment]
Apparently I was a bit rambly this morning.  Sorry :).

We've definitely identified that there's work to be done here, but we don't think the right answer is to bring forward the existing System.Security.Cryptography.Xml code.  Instead, this represents the item on our backlog to engineer a general-purposed implementation for XmlDSig which is fast and not tied to legacy object models (e.g. XmlDocument).

That effort isn't something that we're expecting to do for .NET Core 1.0, simply because we're focused on other things.

But, letting us know that you're impacted by the missing functionality does help with ongoing prioritization.

**Comment by AndersAbel** [comment]
I'm looking at creating an ASP.NET Core SAML2 Middleware, based on https://github.com/KentorIT/authservices. Without a SignedXml port it won't be able to run on the Core framework.

I definitely do agree that not carrying over the existing one is a good idea. Would it be possible to do anything based on an XmlReader API? That way both XDocument and XmlDocument can be supported. Also providing a public implementation of the enveloped reader used in System.IdentityModel would be nice (if it was improved to support XML files with whitespace...)

**Comment by bartonjs** [comment]
@AndersAbel XmlReader is where I'd start, yeah (unless the XML guys tell me there's something better).

Since XmlReader what the System.IdentityModel variant is based on, it should be doable :).

**Comment by AndersAbel** [comment]
@bartonjs The System.IdentityModel variant is quite limited though in what transforms it can handle. For SAML2/WS-Fed work it wouldn't be a problem, but as a general API it must be considered how to deal with non-enveloped signatures and xml containing nested signatures (such as a signed saml response containing signed assertions). Also I think that the System.IdentityModel.EnvelopedSignatureReader is copying the data when doing the validation. There's a lot of fun to be done there. If I had the time I'd love to work on it.

**Comment by Icestorm0141** [comment]
I appreciate the rambling @bartonjs, helps to see whats going on behind the scenes.

Currently my company is impacted by this. We have some legacy code that we're trying to port over to .NET Core that generates signed XML license files and without this set of classes we're stuck. We're open to diverging from XML files as the base for the licenses but as of this moment we haven't found a good solution that fits our needs. 

Looking forward to seeing this get added in the future.

**Comment by sandersaares** [comment]
I can attest that this (and also the slightly related XML Encryption) is relevant for us. The existing form in .NET Framework would be just fine - no need for innovation here, from my perspective. Copy & paste implementation would be very welcome!

Is there currently any workaround available?

**Comment by af0l** [comment]
Would like to know what @sandersaares asked, too. There's no built in way to sign xml in the CoreFX now?

**Comment by bartonjs** [comment]
@sandersaares / @af0l : There is not, for .NET Core 1.0, any inbuilt implementation of SignedXml/XmlDSig.

Based off of comments here (and others) we'll probably just bring over the old API, but we didn't have time to make it happen for 1.0.

**Comment by af0l** [comment]
Thank you @bartonjs, that must be the reason I could not get our project to work on Core. :) It's also a great shame, because I'd love to move ahead and can't until it's done. We have to report all payments to our tax authority using signed xml's, so it's a showstopper really.

**Comment by emailtowalter** [comment]
Any progress on this? I am stuck with the SAML token validation which requires this feature. Thanks

**Comment by af0l** [comment]
Yeah, would liek to know that too. We ended up extracting the features needing the signature and putting them into a separate web API solution...

**Comment by john182** [comment]
Already have idea on which version will be implemented or solution

**Comment by bartonjs** [comment]
At this point it seems that the most straightforward answer is to port the existing .NET Framework implementation to .NET Core.  So we're bucketing this in with other "making it hard to port" API omissions.

**Comment by sandersaares** [comment]
Potentially relevant to the topic: https://connect.microsoft.com/VisualStudio/feedback/details/3002812/xmldsigc14ntransform-incorrectly-strips-whitespace-and-does-it-inconsistently and https://github.com/sandersaares/xml-c14n-whitespace-defect. It would seem to me that the .NET Framework implementation of Canonical XML 1.0 is incorrect. I hope I am wrong but if so, this might introduce some hairy compatibility questions.

**Comment by AndersAbel** [comment]
@sandersaares Looked at your sample and you need to set ` XmlDocument.PreserveWhiteSpace = true` when reading the Xml if it contains whitespace.

**Comment by sandersaares** [comment]
@AndersAbel Thanks for the hint! That changes the situation and actually enables conforming validation if an XML Schema is present. Without an XML Schema, behavior remains invalid (in a new and exciting way). I have updated the Connect issue and GitHub repo accordingly.

**Comment by AndersAbel** [comment]
@bartonjs If you port the existing .NET framework solution over, please fix the bug that prevents verification of multiple signatures: https://connect.microsoft.com/VisualStudio/Feedback/Details/2288620

**Comment by sandersaares** [comment]
FYI if this gets to implementation stage then I have a freshly minted library here, with tests, that makes use of XML signatures (both sign and verify) and other XML functionality on .NET Framework - might be useful to get some dependency-free real-world code to try out the implementation against: https://github.com/Axinom/cpix

**Comment by henkmollema** [comment]
Is there a timeline for development of this API?

//cc @bartonjs

**Comment by bartonjs** [comment]
@henkmollema Nothing specific, no.  For the 1.2 release we're working to shrink the gap between .NET Framework and .NET Core; and that effort is currently where SignedXml is coming from.

**Comment by brockallen** [comment]
I was on a customer call today that needs SAML2-P support on ASP.NET Core (which would be using KentorIT's implementation). This is a blocking issue for customers that want to move to ASP.NET Core. For now my customer will have to stay on Katana.

**Comment by brockallen** [comment]
I see the milestone was changed from 1.2 to Future (by @bartonjs). Can you comment or elaborate?

**Comment by bartonjs** [comment]
Mainly it just has to do with how we're tracking milestones.  We used to do more of "we hope it'll make this one", then at the end of the milestone we'd re-assign everything that wasn't done.  We're now trying really hard to say "if we marked it for this milestone it should be very rare for it to get reassigned".

This is downstream of a lot of other work for the 1.2 milestone, and it was (to me, anyways) always a bit of a stretch for it to make it for 1.2.  It's still pretty high up on our "what's next" list, we're just not 'committing' to it being part of the 1.2 release (which is mainly netstandard2.0 work, bugfixing, and a couple of infrastructure projects).

Marking it as Future doesn't guarantee that it *won't* be part of the 1.2 release, it just means (using our current interpretation of milestone) that we won't hold the release for it.  Once someone is actually working on it then there's a better understanding of when it'll actually get done, and then it'll get marked as whatever milestone it'll actually release in.

@karelz If there's anything you want to add (or correct) feel welcome to chime in.

**Comment by karelz** [comment]
We won't be able to fund the work in 1.2 timeframe (there's just too much other stuff with higher impact to finish). that's why we moved it to Future milestone, to communicate our plans.
We are aware of the number of asks, that's why it is high in our Security area backlog. It is also one of the top asked corefx missing APIs in general (among DirectoryServices, SerialPort, etc.).

cc: @steveharter @danmosemsft @terrajobst

**Comment by karelz** [comment]
Our answers crossed :)
More context on Milestones is in our [issue guide](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#triage-rules---simplified).

The .NET Framework code is available as [reference source](https://referencesource.microsoft.com/#System.Security/system/security/cryptography/xml/signedxml.cs). So technically speaking the port can be initiated even outside of .NET Core team - if folks are interested to help us here.
From my earlier chats with @bartonjs I think the key "challenge" will be creating/porting tests.

**Comment by Jaedson33** [comment]
Hey, what's the actual situation about that problem?

**Comment by karelz** [comment]
@Jaedson33 what do you mean by 'problem'? If you mean when it will be fixed -- see the answers from last week.

**Comment by Jaedson33** [comment]
@karelz  But I don't want to wait. Why you don't fix it now?

**Comment by karelz** [comment]
@Jaedson33 see my reply above - it explains why we cannot fund it now. It is about priorities. There is a line of people who want many features/APIs now, but we do not have infinite-size team, so we have to prioritize.

If you need it really badly ASAP, you can always contribute to the code base, we will be happy to help out with guidance, code reviews and feedback.

**Comment by Jaedson33** [comment]
Ok, so I'll wait.

**Comment by tintoy** [comment]
@karelz if the original tests are still available to verify the work, I'd be willing to put my hand up :)

(one of my coworkers also has relevant experience so we'd probably be working on it together)

**Comment by karelz** [comment]
Key part of the work is actually to create new test assets. The old tests are insufficient and have poor coverage. We will need someone to review the spec and add test for every interesting requirement -- that's where most of the cost is.

If you are still interested, we can try to drop in source code from full .NET Framework as is - the next step will be to get it built and add test coverage, before it can be released as part of .NET Core. Let me know if you're interested ...

**Comment by tintoy** [comment]
Ok, yes - we're still interested :)

**Comment by Jaedson33** [comment]
@tintoy I'm interested to help you because I really need that class.

**Comment by tintoy** [comment]
> @tintoy I'm interested to help you because I really need that class.

Glad to hear it :)

**Comment by Jaedson33** [comment]
So... How can I help?

**Comment by tintoy** [comment]
Let me first have a chat to my coworker and we'll come up with a plan of attack. @karelz - are there any guidelines or other docs we should read before wading in? To start with, I'm guessing my coworker will probably wade into the standard, I'll probably look at working out where the code needs to go (and what's involved in getting existing tests from other parts of the framework to run before we make any changes). Does that sound reasonable?

CC: @anthonylangsworth

**Comment by AndersAbel** [comment]
To keep the scope a bit limited I'd recommend to start without the features that are disabled by MS16-035 (xpath transform, xslt transform, external references). I don't know what room there is for breaking changes, but the current fallback mechanism in DefaultGetIdElement can be exploited in signature wrapping attacks. I'd prefer a more secure default version.

It would also be good if the internal API was restructured a bit to support the EnvelopedSignatureReader used by System.IdentityModel instead of having two separate implementations of XML Signature validation.

Finally I'd also like a single dot to be added as per [this bug report](https://connect.microsoft.com/VisualStudio/Feedback/Details/2288620).

**Comment by karelz** [comment]
@tintoy I don't think we have good docs. I think we should add the sources and then we can parallelize the work - let me sync with @bartonjs @steveharter @ianhays.

**Comment by AndersAbel** [comment]
I'll try to find some time to help too. If there are any questions on the spec and how it works I'll be happy to look into that - I've already spent some time reviewing the spec.

Does anyone has anything to say about the idea to consolidate SignedXml and the EnvelopedSignatureReader used by System.IdentityModel?

**Comment by karelz** [comment]
@AndersAbel

> start without the features that are disabled by MS16-035 (xpath transform, xslt transform, external references)

We should start with latest .NET Framework source code, which should be secure. If you have any concerns about .NET Framework code security, please let us know offline.

>  I don't know what room there is for breaking changes

No room, we should start with straightforward port from .NET Framework. Any further improvements, changes, breaking changes, etc. can be considered later. Not as part of the initial work. Otherwise it will grow over our heads.

> current fallback mechanism in DefaultGetIdElement can be exploited in signature wrapping attacks

That should be treated as separate problem. @bartonjs can you please comment?

> It would also be good if the internal API was restructured a bit to support the EnvelopedSignatureReader used by System.IdentityModel instead of having two separate implementations of XML Signature validation.

Again, let's take it as next step, after we have fully functional port with good test coverage.

> Finally I'd also like a single dot to be added as per this bug report.

Please file it as separate issue here, on GitHub. Ideally after we have the code here ported (i.e. when the bug is truly applicable to .NET Core).

> Does anyone has anything to say about the idea to consolidate SignedXml and the EnvelopedSignatureReader used by System.IdentityModel?

Just want to reiterate. We should treat it as next step, post porting.

**Comment by bartonjs** [comment]
> >    current fallback mechanism in DefaultGetIdElement can be exploited in signature wrapping attacks
>
> That should be treated as separate problem. @bartonjs can you please comment?

@AndersAbel If you believe there is a security concern, please follow the vulnerability reporting process at https://technet.microsoft.com/en-us/security/ff852094.aspx.

> Does anyone has anything to say about the idea to consolidate SignedXml and the EnvelopedSignatureReader used by System.IdentityModel?.

It likely isn't possible.  SignedXml is very heavily (and public-API-ally) based on the rich-DOM XmlDocument.  IdentityModel's representation is based on XmlReader.  But, once the existing stuff is brought over it can be investigated.

**Comment by tintoy** [comment]
> I'll try to find some time to help too. If there are any questions on the spec and how it works I'll be happy to look into that - I've already spent some time reviewing the spec.

@AndersAbel - cheers, I'm sure we could use the help :)

**Comment by AndersAbel** [comment]
@bartonjs I've reported those issues to secure@microsoft.com, which resulted in MS16-035. IMO there are some remaining risky issues though, that MS decided to not fix (they would incur breaking changes). I've not yet published the details, but if you would like to discuss them privately, please mail me.

@karelz Thanks for clearing up that there's no room for breaking changes. That means that my ideas about consolidation are not relevant.

>> start without the features that are disabled by MS16-035 (xpath transform, xslt transform, external references)

> We should start with latest .NET Framework source code, which should be secure. If you have any concerns about .NET Framework code security, please let us know offline.

The MS16-035 patch addressed a number of issues in SignedXml. It is possible though to use registry keys to revert to the old, unsecure behaviour. Should those options also be ported to .NET Core? My suggestion above was meant to prioritise getting the current default .NET Framework behaviour ported, leaving those parts that are deactivated by default behind for now. Or do you mean that those parts are crucial to be moved over too? Then there's the question how to handle the configuration as .NET Core AFAIK doesn't rely on the registry for configuration (as it's not available on all platforms).

**Comment by bartonjs** [comment]
>  It is possible though to use registry keys to revert to the old, unsecure behaviour. Should those options also be ported to .NET Core? 

No.  Registry-compat-only code will be deleted before the package is made available.

**Comment by Jaedson33** [comment]
Why you don't create a project on GitHub to implement that?

**Comment by karelz** [comment]
We synced with @bartonjs, @steveharter and @ianhays

**Goal:** Provide APIs fully compatible with full/Desktop .NET Framework (no changes as part of this work - straight port only)

**Plan:**
1. Add the source code on GH sanitized, with licenses, etc. (it won't build) - @steveharter will do that
2. Make the source code build (still excluded from overall repo build)
3. Remove Desktop registry-compat code paths
    - @steveharter can you please edit this line with additional details how to identify such code?
4. Add tests (We have to agree on expected code coverage and how much of the spec we have to cover)
    - Compare test results between Desktop and .NET Core implementations
5. Make it part of overall repo build & ship it!

Code changes rules: **Only code changes absolutely necessary** to make it build and compatible with (full) .NET Framework are allowed, no additional bug fixes or changing architecture - such PRs will be rejected now.
Changes like that can be considered after the initial port is done, when we have a good test bed and when we are able to validate such change.

If larger code/architecture changes are needed from some reason, we should discuss them first here, before doing the work / submitting PR.

**Comment by tintoy** [comment]
Sounds good to me :)

**Comment by AndersAbel** [comment]
@karelz, @steveharter  Most registry lookups are located in the [Utils](https://referencesource.microsoft.com/#System.Security/system/security/cryptography/xml/utils.cs,4f5d226a87b242bc) class. `AllowAmbiguousReferenceTargets`, `AllowDetachedSignature`, `RequireNCNameIdentifier`. There's also a lookup in the [SignedXml class](https://referencesource.microsoft.com/#System.Security/system/security/cryptography/xml/signedxml.cs,8b616077b30145cd,references) where it setups the list of known transforms. Without the registry compatibility the `XmlDsigXPathTransform` and `XmlDsigXsltTransform` cannot be accessed. Should they be removed completely from the source together with the registry-compat code?

That's those I know about, I've not seen any other while reading the code, but I might have missed something.

**Comment by steveharter** [comment]
@AndersAbel I did update Karel's comment above regarding registry. If there are any classes that are not accessible we need to understand what functionality is being lost. For the ones you mention, I believe CryptoConfig needs to add the name:object pair for those and thus can be created late-bound

**Comment by Jaedson33** [comment]
When do you think this class will be ready?

**Comment by karelz** [comment]
Do you mean for contributions? @steveharter plans to submit the initial "add sources" PR very soon (most likely today).

**Comment by steveharter** [comment]
The initial code has just been merged.

**Comment by Jaedson33** [comment]
@steveharter  Thanks 😃

**Comment by karelz** [comment]
Thanks @steveharter! I have moved the execution plan to the top-most post for easier progress tracking. Whenever we make changes to it, we will mention the changes in another reply here.

**Comment by karelz** [comment]
If anyone wants to start working on it, please say so to avoid duplicated work. We will co-assign the issue to you.

**Comment by anthonylangsworth** [comment]
@karelz: @tintoy and I put our hands up to get started on this. Happy for you to assign it to us.

**Comment by tintoy** [comment]
> If anyone wants to start working on it, please say so to avoid duplicated work. We will co-assign the issue to you.

Cheers - I'm happy to make a start on getting it to compile :)

**Comment by tintoy** [comment]
@anthonylangsworth - hah, beat me to it!

**Comment by tintoy** [comment]
Work's happening [here](https://github.com/tintoy/corefx).

**Comment by karelz** [comment]
Assigned to @tintoy. I will co-assign it also to @anthonylangsworth once he accepts the contributor status to the repo (GH won't offer you as assignee option without it).

**Comment by tintoy** [comment]
Thanks!

**Comment by tintoy** [comment]
@karelz just to confirm - from what I understand of the contributor guidelines I'm to make these changes in `master`?

**Comment by tintoy** [comment]
(my `master` obviously)

**Comment by tintoy** [comment]
Ergh, sorry, let me try again - the eventual PR to be against your `master`?

**Comment by karelz** [comment]
Yes, that's right. Just don't make it part of full corefx build/test run until the very last phase.

**Comment by tintoy** [comment]
I've found a couple of constants that seem to have been moved to an internal class in [src\Common\src\Interop\Windows\Crypt32\Interop.certificates_types.cs](src\Common\src\Interop\Windows\Crypt32\Interop.certificates_types.cs). This isn't accessible from `System.Security.Cryptography.Xml` though. Any thoughts on the best approach here?

**Comment by karelz** [comment]
Which of them are needed? All of them?
Can you check reference source if they are public in .NET Fx? (I don't think so, but doesn't hurt to double check)
I am a bit surprised that we do special interop, instead of leveraging the rest of Crypto library ... either it needs something special, or it is from historical reasons ... @steveharter @bartonjs any thoughts?

**Comment by bartonjs** [comment]
@tintoy One of the things that needs to be done as this effort is removing the direct interfacing with CAPI, switching to using the .NET APIs.

**Comment by tintoy** [comment]
@karelz, @bartonjs - mostly CAPI HRESULT constants passed to the `CryptographicException` constructor.

**Comment by tintoy** [comment]
I'll have a look at how other code in corefx works with `CryptographicException`.

**Comment by tintoy** [comment]
Ah, ok - so it looks like the HResult constructor's not used anymore - just the one that takes a message. I'll see if there are existing message resources that correspond to those `E_xxx` values.

**Comment by tintoy** [comment]
As for the other issues, it looks to me like a result of types not sharing a single assembly anymore. For example, `X509Utils.DecodeHexString` is in `System.Security.Cryptography.X509Certificates` but in the full framework this just lives in the `System.Security` assembly along with the classes that use it.

Since everything's been broken out into multiple assemblies what's your preferred mechanism for dealing with what would ordinarily be shared components? I could just make a copy of the required functions from code in other assemblies if that's what you'd prefer.

**Comment by tintoy** [comment]
Or pull in the source using something like:

```xml
<Compile Include="$(CommonPath)\Interop\Windows\Crypt32\Interop.certificates_types.cs">
    <Link>Interop\Windows\Crypt32\Interop.certificates_types.cs</Link>
</Compile>
```

**Comment by tintoy** [comment]
For now I've worked around the CAPI issue by simply compiling `Interop.certificates_types.cs into the assembly and referencing constants from there.

I also had to copy over some methods from X509Utils.cs in the full framework (mainly to do with Hex encoding / decoding) as there's nothing publicly available in corefx that does this.

The only remaining issues are in [src/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SymmetricKeyWrap.cs (line 34, among others)](https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SymmetricKeyWrap.cs#L34) which result in errors such as:

```
error CA5350: TripleDESKeyWrapEncrypt uses a weak cryptographic algorithm TripleDES
```

For now, I've suppressed error. So now it all all compiles :)

Ok, well except for the test project:
```
corefx\Tools\PackageLibs.targets(34,5): error : Could not locate compile assets for any of the frameworks .NETStandard,Version=v1.7
corefx\Tools\PackageLibs.targets(34,5): error : Could not locate runtime assets for any of the frameworks .NETCoreApp,Version=v1.1
```
I'll work that out tomorrow.

**Comment by tintoy** [comment]
@karelz @bartonjs I'm going to open a PR so we can discuss the changes (the work is basically done as far as I can tell) - are you OK with that?

**Comment by karelz** [comment]
Sounds good to me. @steveharter any thoughts?

**Comment by Jaedson33** [comment]
Happy new year =D

**Comment by Jaedson33** [comment]
Do you have any idea when phase 2 will be completed?

**Comment by karelz** [comment]
#14662 is already merged - that was the phase 2. I will mark it in the top post as 'checked'.

**Comment by brockallen** [comment]
Just so I'm clear: once all 5 steps are done, then to get ws-fed support in ASP.NET Core, the AAD team needs to do their SAML token bits, then the ASP.NET teams needs to build the ws-fed middleware on the AAD piece. Does that match your expectations?

**Comment by karelz** [comment]
Nope, this work has nothing to do with WS-Fed.
I owe a reply and explanation on https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/500

**Comment by karelz** [comment]
FYI: Reply posted: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/500#issuecomment-275218749

**Comment by Jaedson33** [comment]
So when phase 4 will be completed?

**Comment by karelz** [comment]
If you're asking when .NET team will have time to get to it -- we don't know yet. It is high on our backlog, but there are couple of higher priorities we have to tackle first.

We are open to contributions in the space from the community in the meantime.

**Comment by tintoy** [comment]
I'm happy to keep working on this, but I'm kinda stuck at the moment; since corefx moved to the new build process, `System.Security.Cryptography.Xml` no longer builds (so @anthonylangsworth and I are blocked from writing any tests). If we could get just a quick pointer on what's involved in getting the project (and test project) to build, we'll be good to go :)

PS. I've spent 20 minutes or so tracing through the build process to work out why it no longer builds, but haven't worked that out yet. Any pointers would be appreciated...

**Comment by karelz** [comment]
@mellinoe @weshaggard can you please provide guidance for migrating SignedXml to the new build system?

**Comment by Jaedson33** [comment]
😭😭 I think I'll need to wait 😭😭

**Comment by weshaggard** [comment]
@tintoy if you point me at the branch you are currently working on and which project you are trying to build I can help with getting building.

**Comment by karelz** [comment]
@weshaggard there is currently not a branch - the code is in master, it just is not hooked up into root build (on purpose) - src/System.Security.Cryptography.Xml (introduced in #14628). @steveharter can provide additional info.

**Comment by tintoy** [comment]
@weshaggard @karelz I'm happy to create a branch in our fork and only get it building there to unblock us; later, we can always cherry-pick any changes we've made once it's building again in `master`. Let me know if this is your preferred approach :)

**Comment by weshaggard** [comment]
PR https://github.com/dotnet/corefx/pull/15491 should get the project infrastructure working. Once CI passes for it we can merge it and it should bootstrap your efforts.

**Comment by tintoy** [comment]
Thanks - I've rebased [tintoy/corefx/master](/tintoy/corefx/tree/master) and will have a play with it shortly :)

**Comment by tintoy** [comment]
@weshaggard ok, so both src and test project build, but if I add a project reference from the test project to the source project (`<ItemGroup><ProjectReference Include="..\src\System.Security.Cryptography.Xml.csproj" /></ItemGroup>`), I get:

```
1>------ Build started: Project: System.Security.Cryptography.Xml, Configuration: Debug Any CPU ------
1>  D:\Development\github\tintoy\corefx\src\System.Security.Cryptography.Xml\src\System.Security.Cryptography.Xml.csproj ConfigurationErrorMessage: Could not find a value for TargetGroup from Configuration 'Debug'.
1>  System.Security.Cryptography.Xml -> D:\Development\github\tintoy\corefx\bin\AnyOS.AnyCPU.Debug\System.Security.Cryptography.Xml\netcoreapp\System.Security.Cryptography.Xml.dll
2>------ Build started: Project: System.Security.Cryptography.Xml.Tests, Configuration: Debug Any CPU ------
2>D:\Development\github\tintoy\corefx\buildvertical.targets(88,5): error MSB4018: The "FindBestConfiguration" task failed unexpectedly.
2>D:\Development\github\tintoy\corefx\buildvertical.targets(88,5): error MSB4018: System.ArgumentException: Property 'ConfigurationGroup' value 'Debug' occured at unexpected position in configuration 'Debug'
2>D:\Development\github\tintoy\corefx\buildvertical.targets(88,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.ConfigurationFactory.ParseConfiguration(String configurationString, Boolean permitUnknownValues) in D:\Development\github\tintoy\corefx\src\Tools\CoreFx.Tools\Configuration\ConfigurationFactory.cs:line 219
2>D:\Development\github\tintoy\corefx\buildvertical.targets(88,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.FindBestConfiguration.Execute() in D:\Development\github\tintoy\corefx\src\Tools\CoreFx.Tools\FindBestConfiguration.cs:line 34
2>D:\Development\github\tintoy\corefx\buildvertical.targets(88,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
2>D:\Development\github\tintoy\corefx\buildvertical.targets(88,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
========== Build: 1 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

I'm guessing I've missed something about how inter-project references are meant to work in corefx?

**Comment by weshaggard** [comment]
@tintoy I'm not sure what that error is specifically but we generally don't need ProjectReferences in our new engineering system. For the test build we always build and reference the full targeting pack, in this case produced in `bin\ref\netcoreapp`. The library that is put into that directory is what you build from your ref folder which is currently empty. So to get yourself going you need to generate a ref with the API surface area that you need and get the ref building, then your test project will automatically see the APIs and can build against them. We have a tool called genapi that can generate the ref based on another library. Let me submit another PR quickly to seed that for you guys.

**Comment by tintoy** [comment]
Ok, now I get it now; the reason I see no types in the test project is because the ref project has no types yet :-)

**Comment by tintoy** [comment]
@weshaggard if you don't have time, I can probably work out how to do that; I was just reading about that tool the other day.

**Comment by weshaggard** [comment]
I quickly ran the genapi tool and pushed that commit https://github.com/weshaggard/corefx/commit/29cf289f3e007fd4b13b191866ae848d99dec67e. Feel free to take that or regenerate it yourself. You will need to add ProjectReference's to other refs to get it to compile.

**Comment by tintoy** [comment]
Cheers - that'll save me time, much appreciated.

**Comment by tintoy** [comment]
@weshaggard success! Thanks, hopefully we can get onto writing those tests now :)

**Comment by tintoy** [comment]
(tintoy@dd834c63af4fe40faf84bc6a776b474ec9947eb1, just ignore the duplicate)

**Comment by tintoy** [comment]
Yep! Have a working test:

https://github.com/tintoy/corefx/commit/24864b3d25e665b6305f116890328527db07f1e1

Thanks for your help, @weshaggard!

PS. I had to locally override (via `.user` file) the executable path under the test project properties (Debug tab). Was `D:\Development\github\tintoy\corefx\Tools/testdotnetcli/dotnet.exe` but only worked when I changed it to `D:\Development\github\tintoy\corefx\Tools\testdotnetcli\dotnet.exe`.

**Comment by weshaggard** [comment]
>D:\Development\github\tintoy\corefx\Tools/testdotnetcli/dotnet.exe

My VS doesn't complain about the slashes. The path separators are generally a pain because we are trying to make things work on both windows and unix, so we end up seeing a bunch of mixed slashes on windows which is generally more accepting then unix.

**Comment by tintoy** [comment]
Just out of curiosity, which version of VS are you using? 2015 or 2017? Perhaps it's been fixed in 2017 :)

**Comment by tintoy** [comment]
(I'm usually on OSX or Linux these days, so I totally appreciate the effort to make things cross-platform friendly, BTW)

**Comment by weshaggard** [comment]
I'm using VS 2015 on Windows 10 and opened the solution and could F5 to debug the tests and my debug path is `D:\git\corefx\Tools/testdotnetcli\dotnet.exe`

**Comment by tintoy** [comment]
Ok, I must be going crazy - now I can't reproduce it either!

Beforem it complained about not being able to find `dotnet.exe`, but it started working when I explicitly set the executable to a path with backslashes only. I just removed the `.csproj.user` file and it _still_ works, so who knows :-o

**Comment by solomon87** [comment]
Hey guys, I'd love to get involved in helping to write tests. What can I do to get started contributing?

**Comment by tintoy** [comment]
Great - I think we might be at the stage now where we can start doing that :)

Let me check with a coworker to see if he's successfully run his first test...

cc: @anthonylangsworth

**Comment by tintoy** [comment]
@karelz,

@anthonylangsworth and I have [started writing tests](https://github.com/tintoy/corefx/tree/feature/xml-crypto/tests/src/System.Security.Cryptography.Xml/tests); I'm aware that we haven't reached common agreement on what needs to be tested but we're going to kick off anyway and then port the tests over to wherever we agree to do the work.

**Comment by tintoy** [comment]
And for those that have volunteered to help with the tests (which is hugely appreciated), we should be in a position to divide up the suggested work early next week :)

**Comment by karelz** [comment]
Regarding test coverage and which tests we need - @bartonjs had strong opinion on it (as he has most expertise with the type and its troubles on our side). He suggested to write tests by the spec (I mentioned it [above](https://github.com/dotnet/corefx/issues/4278#issuecomment-266604862)).
He will be back from vacation at the end of next week. We should probably discuss details & expectations with him.
Also @AndersAbel mentioned spec expertise and potential help [earlier in the discussion](https://github.com/dotnet/corefx/issues/4278#issuecomment-266786882).

cc @steveharter if he has additional guidance while @bartonjs is OOF.

Thanks @tintoy @anthonylangsworth for your contributions here! We really appreciate it!

**Comment by jnm2** [comment]
> cc @**steveharter** if he has additional guidance while @**bartonjs** is OOF.

My curiosity reached a point which demanded to be satisfied.
https://blogs.technet.microsoft.com/exchange/2004/07/12/why-is-oof-an-oof-and-not-an-ooo/
I feel better now.

**Comment by karelz** [comment]
😃 I didn't even know it's so Microsoft specific! Thanks for the enlightenment 😃

**Comment by tintoy** [comment]
@anthonylangsworth has started sketching out a rough test plan so we at least have something to discuss. Feel free to have a look over it and provide feedback :)

CC: @karelz @steveharter @bartonjs

**Comment by anthonylangsworth** [comment]
@karelz @steveharter @bartonjs (or anyone it may concern) What is the policy around using the `InternalsVisibleToAttribute` for tests? There are a lot of `internal` classes within that namespace and getting adequate test coverage may be hard going solely via public methods. However, I appreciate there are other considerations.

**Comment by karelz** [comment]
Hmmm, sadly, I don't know - @weshaggard @stephentoub? (question in previous reply)

**Comment by tintoy** [comment]
From what I've seen of other code in corefx, the projects in question sometimes include the relevant source files from other projects (although I imagine this will get complicated pretty quickly due to dependency graphs).

**Comment by tintoy** [comment]
From memory, `System.Collections.Immutable` uses `[InternalsVisibleTo]`, if that's any help.

**Comment by tintoy** [comment]
For example:

https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/Properties/InternalsVisibleTo.cs

**Comment by tintoy** [comment]
More generally:

https://github.com/dotnet/corefx/search?utf8=%E2%9C%93&q=%22assembly%3A+InternalsVisibleTo%22

**Comment by weshaggard** [comment]
You can see my thoughts on InternalsVisibleTo in this old issue https://github.com/dotnet/corefx/issues/1604.  My general take is not to do it unless there is a real specific need to do it.

**Comment by bartonjs** [comment]
>  What is the policy around using the InternalsVisibleToAttribute for tests?

No.

> There are a lot of internal classes within that namespace and getting adequate test coverage may be hard going solely via public methods. 

If it's something other than a deep exception path it should be reachable, or deleted.  If it takes a tricky test case to hit it, well, that's what we need.

> projects in question sometimes include the relevant source files from other projects

Assuming you mean "the test projects include the product source to get access to internal members": No.

-----

`[InternalsVisibleTo]` and source sharing are both inherited strategies.  The most vociferous of us believe (essentially) that our tests should only be representative of things that could be encountered in the real world.  If no test is possible to hit some particular block then why does it exist?  Yes, there are some exception-throwing blocks that can't be hit because a redundant check higher in the stack already caught it; but that is something that can be looked at after the fact and declared OK.

So I'd recommend pulling up [the spec](https://www.w3.org/TR/2002/REC-xmldsig-core-20020212/)  and ensuring that there is a positive test for every feature (e.g. every algorithm it says it supports, and boundary cases for min/max values for options on those algorithms).

Negative tests like removing required elements (e.g. 4.1 says that `SignedInfo` is a required element for `Signature`... do we emit a reasonable exception if it's missing?) are also good.

Canonicalizer option tests, like `<foo><!-- hi --></foo>` and `<foo></foo>` (and maybe `<foo />`?) are the same under `c14n`, but different under `c14n-withcomments`.  (This probably requires signing both ways, then swapping bodies, since the canonicalization algorithm should be signed).

Transform tests.  All canonicalizers are transforms.  Etc.

Tamper tests are also good.  But if you believe you found a tamper test that successfully tampers, don't report it here or commit/push it to any repro on github (email the test/case/description to secure@microsoft.com).

---

Okay, I've thought too much for the day.  Back on vacation now.

**Comment by karelz** [comment]
Thanks @bartonjs for your insights while on vacation! Now go and have some fun in real world ;-)

**Comment by tintoy** [comment]
@karelz @bartonjs (although only once you're back from vacation!) @steveharter and anybody else with an opinion:

@anthonylangsworth has created [some initial tests](https://github.com/tintoy/corefx/tree/feature/xml-crypto/tests/src/System.Security.Cryptography.Xml/tests) as a discussion-starter, and we'd appreciate any feedback you may have so far.

**Comment by steveharter** [comment]
I see that Mono has some SignedXml tests. These should be evaluated and vetted against the xmldigsig spec , suggestions that @bartonjs mentioned earlier, and current code to see what\if they are worthwhile to port\bring over. Ping me for copyright (header) information if these tests make sense to bring over.

**Comment by tintoy** [comment]
Thanks - we'll have a look at that :)

**Comment by anthonylangsworth** [comment]
Thanks for that, @steveharter . Can you provide links, please? These could short cut a lot of our testing. Are there any copyright or other considerations if we expand or build on these instead of copying them verbatim?

**Comment by karelz** [comment]
@anthonylangsworth when copying source code from Mono, we have to keep & update the copyright header. I would suggest to first do just a copy (with the right headers, potentially minor tweaks). Once we have the code in CoreFX, we can modify the code as we wish.

**Comment by steveharter** [comment]
@anthonylangsworth Mono tests - https://github.com/mono/mono/blob/0bcbe39b148bb498742fc68416f8293ccd350fb6/mcs/class/System.Security/Test/System.Security.Cryptography.Xml/SignedXmlTest.cs

**Comment by anthonylangsworth** [comment]
Thanks, @steveharter. I have started converting the tests over from NUnit to Xunit.

**Comment by danmosemsft** [comment]
I posted this but realized it's in @anthonylangsworth 's repo:

Here's the magic to do to the copyright headers when we pull code from Mono:

	1. Keep the existing copyright headers in place
		○ Note: If you are porting just portions of the file over, you still need to bring over all the existing copyright headers.
	2. Add the following copyright headers – NOTE this is missing the middle line of the ‘regular’ CoreFx copyright headers:
                 // Licensed to the .NET Foundation under one or more agreements.
                 // See the LICENSE file in the project root for more information.

**Comment by Jaedson33** [comment]
Are there any failing tests in this project?

**Comment by anthonylangsworth** [comment]
@Jaedson33 We are currently converting the mono tests. I have not found any failing tests that indicate code errors yet but we still have many tests to go.

**Comment by Jaedson33** [comment]
@anthonylangsworth What can I do to help?

**Comment by anthonylangsworth** [comment]
@Jaedson33 I have answered that question at https://github.com/tintoy/corefx/issues/6#issuecomment-280904587 . To summarize, we have 84 mono tests still failing (mainly due to changed defaults and .Net core limitations). Any help to get the remaining tests working is appreciated. Otherwise, I'm working through them.

**Comment by anthonylangsworth** [comment]
@karelz @bartonjs @steveharter  The [System.Security.Cryptography.CryptoConfig class](https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs) class states that many XML transforms are not supported on CoreFx (lines 281 to 303 at the bottom of `DefaultNameHT`). 

These correspond to URIs used by classes in the `System.Security.Cryptography.Xml` namespace. As part of adding `System.Security.Cryptography.Xml` back into .Net Core, I assume we should reinstate these. Please let me know if I am mistaken.

CC: @tintoy

**Comment by morganbr** [comment]
@anthonylangsworth , some of those transforms, such as XmlDsigC14NTransform are fine, but others, such as XmlDsigXsltTransform are considered very dangerous. While you can enable them via registry key opt-in in the full .NET Framework, I'd rather not support them on .Net Core. Take a look at KnownCanonicalizationMethods and DefaultSafeTransformMethods at https://github.com/dotnet/corefx/blob/ac17228d823a07a15fe53069a49fb5e5f35835b7/src/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs#L741 for the transforms that we should support.

**Comment by morganbr** [comment]
I hadn't realized the source of dangerous ones actually got included in the port. I'd vote for removing them entirely. There's really no safe way to use them.

**Comment by anthonylangsworth** [comment]
@morganbr Thanks for the heads up. Once I get the RSA and DSA key value XML code changes done, I'll review the list of transforms included. I will likely publish a list and ask you and others to review it.

**Comment by morganbr** [comment]
@anthonylangsworth Sounds great!

**Comment by AndersAbel** [comment]
@morganbr @AnthonyDGreen Those transforms (that was disabled in the MS16-035 patch) have been discussed previously in this thread when discussing the port. @bartonjs stated on Dec 14 that reg-compat stuff should be removed. See also comment by @steveharter Dec 15 that those transforms probably can be enabled late-bound and should thus be ported.

**Comment by karelz** [comment]
@steveharter @bartonjs can you please chime in?

**Comment by steveharter** [comment]
Even with no registry support, CryptoConfig can creates those transforms late-bound by string name or oid. Search for CryptoConfig in the SignedXml code as it is used to create the appropriate class based on the xml content.

This means the CryptoConfig class should be extended to support these transforms, at least for the same types in netfx anyway, and ideally also cross-reference those against Mono. That is unless there is a reason (that I'm not aware of) to not include them and not hook them up to CryptoConfig..

FWIW here's the netfx version of CryptoConfig (in order to compare with what's in corefx): https://referencesource.microsoft.com/#mscorlib/system/security/cryptography/cryptoconfig.cs,20d26e036bc718bc

**Comment by bartonjs** [comment]
There's a list of "allowed transforms" which is (in .NET Framework) registry-extensible for back-compat.  For .NET Core it's not extensible, but hard coded to include only the no-input transforms.

Since you can't validate a document using transforms not on the allow-list, maybe it doesn't make sense to port them.  But since someone *could* be using the transforms outside of the context of SIgnedXml (just wanting to use them as a general purpose transform engine) maybe it's fine to have them.

Since we'd be talking about removing a whole type it wouldn't create an inconsistency with .NET Framework... so I'd probably advocate removing any transforms not already on the hard-coded allow list.  "Remove before publishing the package" can be changed later.  "Publish them" cannot.

**Comment by anthonylangsworth** [comment]
@bartonjs beat me to it. `CryptoConfig` contains a list of allowed transforms. I had to modify that class to allow the transforms in the new namespace, `System.Security.Cryptography.Xml`. While some of the allowed transforms use a .Net class's full name, `CryptoConfig` still only permits a fixed list.

I would prefer to not port transforms that have known security issues but backwards compatibility is also important. Should we be making this decision on behalf of the developer? Do we modify `CryptoConfig` to allow some form of extension so a developer can add new transforms if they choose? How do we reach a decision on this?

**Comment by bartonjs** [comment]
CryptoConfig isn't the limiting factor: https://github.com/dotnet/corefx/blob/ac17228d823a07a15fe53069a49fb5e5f35835b7/src/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs#L763-L787.  (Though, if CryptoConfig is still being used as the factory for resolution then any transform would need to be in both places)

The types providing algorithms not in that list (which are not also canonicalizers) effectively have no value.

Allowing extension to the safe list would require new API, so is technically beyond the scope of this effort.

I'd, personally, leave out the transform types that can't be used in document verification.  They're going to be hard to test.

**Comment by anthonylangsworth** [comment]
Actually, both `CryptoConfig` and `DefaultSafeTransformMethods` are relevant. `SignatureDescription` creation is done in `CryptoConfig` so, irrespective of the values in `DefaultSafeTransformMethods`, you cannot use a transform if it is not mentioned in `CryptoConfig`. `DefaultSafeTransformMethods` restricts that list so, if a transform is specified in XML but not returned by `DefaultSafeTransformMethods`, `SignedXml.CheckSignature` returns false.

After my previous comment, I found the property [`SignedXml.SignatureFormatValidator`](https://msdn.microsoft.com/en-us/library/system.security.cryptography.xml.signedxml.signatureformatvalidator(v=vs.110).aspx). This allows a user to specify what transforms and hash algorithms are appropriate. The caller can use this to override `DefaultSafeTransformMethods`, for example.

While this is a step in the right direction, I would prefer a simple list of editable transforms and hash algorithms. This allows developers to remove unsafe ones or add their own. However, this would involve modifying the interface.

As I asked before, who makes the decision here? As a "security guy", my preference would be to exclude insecure options but potentially at the expense of backwards compatibility.

**Comment by karelz** [comment]
@anthonylangsworth this discussion is part of the decision making. Area experts with the most background will make the decision here.

My recommendation: If it is questionable what is the right action, I would suggest to keep status quo - leave the code as is during the port exercise (potentially even without any test coverage) and decide later, separately from the port exercise.

**Comment by bartonjs** [comment]
Okay, so I chatted with some people internally and I think I have the scaffold of a plan:

* Leave the types as public.  (They're public types, and removing things makes people sad).
* These can't (yet) be used in end-to-end SignedXml tests.  (In fact, negative tests seem good)
* They do have public members which should be sufficient for unit testing, so we should do that.
  * And that goes for the rest of the transform types, too.
* If, for some reason, the input-accepting transforms don't work, and everything else is fine, we can figure out what to do about them.  (This would have also applied to "if they had complex compilation dependencies that can't be met", but we're already past that hurdle).

And if/when API or other configuration comes along to enable these types to work, adding the opted-in E2E tests will be easy.

**Comment by krwq** [comment]
@tintoy @anthonylangsworth @peterwurzinger what's the status of your test branch? can we start merging it? I can cherry pick your tests and keep going from there.

Could you also PTAL at https://github.com/dotnet/corefx/pull/16545 ? I've enabled one of the MSDN samples

**Comment by tintoy** [comment]
Hi @krwq - let me have a quick chat to @anthonylangsworth and get back to you.

**Comment by tintoy** [comment]
PS. #16545 -> :+1:

**Comment by tintoy** [comment]
@krwq - had a quick chat to @anthonylangsworth. He mentioned (and I agree) that given the amount of work that's been done there perhaps it'd be good to merge what we have before going any further. It builds, and most tests pass, but a few do not.

I suspect we'll need to rebase against #16545 (which I'll do) before opening a PR (which @anthonylangsworth will do).

We'll get back to you as soon as we're ready for this.

**Comment by anthonylangsworth** [comment]
@krwq To expand on what @tintoy said, while there is still some work to go, a lot of work also has been done porting existing tests and expanding them. In particular, I have already changed `CryptoConfig.cs` to handle many of the algorithms you mention. We all want to move this forward. We also do not want to duplicate or overwrite each other's work. Therefore, we plan to merge the changes as is so others can build on top of the work we have done, find all our bugs and, hopefully, get everything into master sooner.

**Comment by krwq** [comment]
@tintoy is https://github.com/tintoy/corefx/tree/feature/xml-crypto/tests the only branch you are working on?

**Comment by tintoy** [comment]
Yep.

**Comment by peterwurzinger** [comment]
@krwq Well, there is a [PR](https://github.com/tintoy/corefx/pull/7) with a good amount of stuff from Mono in (which is also a sub-branch from  https://github.com/tintoy/corefx/tree/feature/xml-crypto/tests ...). If you're interested in up-to-date code you should probably have a look there. As far as I can tell there are ~40/340 tests failing.

**Comment by tintoy** [comment]
Good catch, @peterwurzinger, I thought we'd already merged that one!

**Comment by krwq** [comment]
@peterwurzinger this PR is pretty nice! I have in fact missed it. Will you be merging that to your branch, to corefx or do you want me to pick it up and do all the merges/rebases? Does it miss anything from Mono tests or is it a complete? @anthonylangsworth - for the CryptoConfig changes - I talked about that with @bartonjs - we should be not touching that file - it's already ugly enough.

My original plan was to get few samples from MSDN and make them all work so that we can start early dogfooding. After that I'll be merging and fixing tests from your branch. If you got some tests failing we can simply disable them for now and fix them later. Let's merge your stuff into corefx/master ASAP :)

**Comment by anthonylangsworth** [comment]
Thanks for the update, @krwq . If you can, please keep us up to date. It helps to know what we are aiming at.

**Comment by Jaedson33** [comment]
Hi =D

**Comment by Jaedson33** [comment]
Are there any failing tests?

**Comment by krwq** [comment]
@Jaedson33 @anthonylangsworth @tintoy 
Here is the current list of (arbitrarily) most important issues:
https://github.com/dotnet/corefx/issues?q=is%3Aopen+is%3Aissue+label%3ASystem.Security.Cryptography.Xml+label%3A%22up+for+grabs%22

**Comment by Jaedson33** [comment]
Hi, do you have any idea when this will be without any error?

**Comment by krwq** [comment]
@Jaedson33 every piece of software has errors :wink: Are there any particular things which doesn't work you?

**Comment by JaedsonBarbosa** [comment]
It's simple: It just don't work in my UWP project 👌

**Comment by JaedsonBarbosa** [comment]
Hi, I'm getting this error when I try to run build.cml. Can you help me?

Error in the command: 
C:\Users\jaeds\Source\Tools\msbuild.cmd /nologo /verbosity:minimal /clp:Summary /maxcpucount /nodeReuse:false /l:BinClashLogger,Tools\net46\Microsoft.DotNet.Build.Tasks.dll;LogFile=binclash.log /p:ConfigurationGroup=Debug /p:BuildPackages=false  /flp:v=normal  /flp2:warningsonly;logfile=msbuild.wrn  /flp3:errorsonly;logfile=msbuild.err  C:/Users/jaeds/Source/Repos/corefx/src/Native/../../bin/obj/Windows_NT.x64.Debug/native\install.vcxproj /t:rebuild /p:Configuration=Debug /p:Platform=x64 /p:PlatformToolset=v141. => O sistema n├úo pode encontrar o arquivo especificado
Command execution failed with exit code 1.
Failed to generate native component build project!
Command execution failed with exit code 1.

**Comment by krwq** [comment]
@JaedsonBarbosa Are you running tests from the developer command prompt? (btw this is slightly OT) for the UWP - I will be investigating if this can be done fairly cheap for 2.0 although no promises

**Comment by JaedsonBarbosa** [comment]
I'm doing that from the Developer Command Prompt for Visual Studio 2017

**Comment by krwq** [comment]
@JaedsonBarbosa have you have pulled the latest changes from github and try building after cleaning your repo (`git clean -fdx`)? Second thing to try is to reduce the path length (i.e. put your repo under C:\corefx\). Another thing to try is to clean nuget cache (%USERPROFILE%\AppData\Local\NuGet and %USERPROFILE%\.nuget)

If still happening please create a separate issue with information:
- your OS
- your VS version
- exact steps you did
- full log (if big put it on gist)

**Comment by JaedsonBarbosa** [comment]
I think it occurs because I think the path C:/corefx/src/Native/../../bin/obj/Windows_NT.x64.Debug/native\install.vcxproj isn't valid.

OS: Windows 10
VS: 2017 Community
I just start the Developer Command Prompt for VS 2017 and put:
```
cd C:/corefx
build
```

**Comment by JaedsonBarbosa** [comment]
I give up, I can't make it work with the 2017 VS.
So do you have any idea of when this may be downloaded as a NuGet package?

**Comment by karelz** [comment]
You should be able to consume the preview from myget.org.
Official package will be part of the .NET Core 2.0 wave - see [milestone 2.0.0](https://github.com/dotnet/corefx/milestone/4) description, May 10 is the ZBB (#17619), so the RTW release should follow "shortly" after it (exact dates are not yet publicly disclosed)

**Comment by JaedsonBarbosa** [comment]
@karelz  Can you please send me the link of the package that contains the System.Security.Cryptography.Xml?

**Comment by karelz** [comment]
https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Security.Cryptography.Xml

Here's how to dogfood 2.0: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md

**Comment by JaedsonBarbosa** [comment]
OK, I want do install that in a UWP class library, but when I try that I get that error:
O pacote System.Security.Cryptography.Xml 4.4.0-preview1-25205-01 não é compatível com uap10.0 (UAP,Version=v10.0). O pacote System.Security.Cryptography.Xml 
4.4.0-preview1-25205-01 dá suporte a:
  - monoandroid10 (MonoAndroid,Version=v1.0)
  - monotouch10 (MonoTouch,Version=v1.0)
  - netcoreapp2.0 (.NETCoreApp,Version=v2.0)
  - uap10.1 (UAP,Version=v10.1)
  - xamarinios10 (Xamarin.iOS,Version=v1.0)
  - xamarinmac20 (Xamarin.Mac,Version=v2.0)
  - xamarintvos10 (Xamarin.TVOS,Version=v1.0)
  - xamarinwatchos10 (Xamarin.WatchOS,Version=v1.0)

**Comment by krwq** [comment]
@JaedsonBarbosa, we currently do not support UAP for that library, you'll need to wait until https://github.com/dotnet/corefx/pull/17969 is merged and new package produced.

**Comment by JaedsonBarbosa** [comment]
😧 Ok, I'll continue waiting 😭

**Comment by JaedsonBarbosa** [comment]
@krwq  Why you don't merge the PR #17969 now?

**Comment by krwq** [comment]
@JaedsonBarbosa It just got merged, I believe the package should be there tomorrow morning - we usually don't merge unless the PR is green on the CI - we got some OSX CI issues since yesterday so it got stale a little bit

**Comment by JaedsonBarbosa** [comment]
@krwq Can you let me know when the NuGet package with the changes can be downloaded?

**Comment by karelz** [comment]
@JaedsonBarbosa be warned that .NET Standard 2.0 support in UWP is bleeding edge -- it won't be part of .NET Core 2.0 -- it will be part of the next version, temporarily named as 2.1. We are working on some of the 2.1 stuff ahead of time and in parallel with 2.0, to stand up infrastructure, etc., so that we can then heavily parallelize (after May 10 which is our ZBB for 2.0).
Either way, you might encounter some issues in using the library in UWP. We might not be in position to help you out right away. We should be able to help more after May 10. ... just setting the expectations.

**Comment by krwq** [comment]
@JaedsonBarbosa seems like we didn't produce any new version for 2 days 😞 You can observe changes here:
https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Security.Cryptography.Xml
whenever a package appears after 4/6 it should have the change you need. I'll check today why there was no package - it could be related to networking issues we have with OSX builds

**Comment by JaedsonBarbosa** [comment]
@krwq Do you now if today the OSX network issues will be solved?

**Comment by danmosemsft** [comment]
It is blocking almost all our PR's so its high priority but we don't have an ETA

**Comment by JaedsonBarbosa** [comment]
OK 👍
Good luck
Buena suerte
Boa sorte

**Comment by JaedsonBarbosa** [comment]
When the part 4 will be ready?

**Comment by krwq** [comment]
@JaedsonBarbosa we have most significant parts already tested and proved working, With testing there is no single "finished" point - you can have 100% code coverage and code which doesn't work. Are there any particular scenarios you are interested in?

**Comment by JaedsonBarbosa** [comment]
No 😃

**Comment by karelz** [comment]
@krwq I expect we will declare things done for the library, when we are comfortable shipping it as stable. When we do that, we can check the box and close this issue. So how far are we test-wise?

**Comment by krwq** [comment]
@karelz I'm currently comfortable shipping as is since all important E2E scenarios I could think of are tested. I still would like to increase coverage so that less popular scenarios (including error handling) can be proven working correctly too although I do not expect finding any 2.0 blocking issues.

For me "done" means that no one will maintain or improve the library anymore

**Comment by karelz** [comment]
OK, if @bartonjs agrees with ready-to-ship-state, then let's be practical and close this issue.
If we want to increase test coverage (as non blocking 2.0), we should create separate work item for Future.

**Comment by bartonjs** [comment]
If we've hit everything off of the list of algorithms and transforms and canonicalizers, it's good with me.

**Comment by JaedsonBarbosa** [comment]
So I think this issue will finally be closed.

**Comment by krwq** [comment]
Ok, let's track the coverage progress with: https://github.com/dotnet/corefx/issues/16829 - I thought we treat it as a master issue

**Comment by karelz** [comment]
Yes, we treat it as master issue, but sometimes you have to declare things done, otherwise you would have at 1 open issue for every larger feature tracking "do more" - which is not truly helping anyone.

**Comment by JaedsonBarbosa** [comment]
It's now closed 😢 
So... Where should I ask questions about the System.Security.Cryptography.Xml?

**Comment by karelz** [comment]
Isn't @krwq's answer above sufficient?
If you have bigger questions, file a new issue. If it is small clarificiation of previous answer, you can keep it here. If not sure, ask here and in the worst case we will ask you to file a new issue ;-)

**Comment by JaedsonBarbosa** [comment]
@krwq  It continue without support for UWP 😞

**Comment by krwq** [comment]
@JaedsonBarbosa does https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Security.Cryptography.Xml/4.4.0-preview1-25210-01 not work for you? Could you send the steps of what you're doing?

**Comment by JaedsonBarbosa** [comment]
I just put that command:
Install-Package System.Security.Cryptography.Xml -Version 4.4.0-preview1-25210-01

That's the error:

> O pacote System.Security.Cryptography.Xml 4.4.0-preview1-25210-01 não é compatível com uap10.0 (UAP,Version=v10.0). O pacote System.Security.Cryptography.Xml 
> 4.4.0-preview1-25210-01 dá suporte a:
>   - monoandroid10 (MonoAndroid,Version=v1.0)
>   - monotouch10 (MonoTouch,Version=v1.0)
>   - netcoreapp2.0 (.NETCoreApp,Version=v2.0)
>   - uap10.1 (UAP,Version=v10.1)
>   - xamarinios10 (Xamarin.iOS,Version=v1.0)
>   - xamarinmac20 (Xamarin.Mac,Version=v2.0)
>   - xamarintvos10 (Xamarin.TVOS,Version=v1.0)
>   - xamarinwatchos10 (Xamarin.WatchOS,Version=v1.0)
>

**Comment by JaedsonBarbosa** [comment]
@krwq  See that: https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Security.Cryptography.Xml/4.4.0-preview1-25210-01#targetframeworks

**Comment by karelz** [comment]
Just a reminder of what I said earlier: https://github.com/dotnet/corefx/issues/4278#issuecomment-292448824
I don't think you will get UWP support with the package. The package depends on .NET Standard 2.0 AFAIK and UWP does not yet have .NET Standard 2.0 support -- it is something we will work on for .NET Core 2.1 (some bits are done early to unblock larger team to work on it post 5/10, but it is not fully functional).
IMO to get UWP support for the package you will have to wait on 2.1.

**Comment by JaedsonBarbosa** [comment]
@karelz  So do you think I will need to wait until when?

**Comment by karelz** [comment]
Yes, I think you will need to wait until then.
Unless you are super-motivated and can make it through all these speed bumps. As I said, until 5/10 our ability to help you with anything UWP will be **very limited**, so you will be mostly on your own 😦. Just setting the expectations here ...

**Comment by JaedsonBarbosa** [comment]
OK, so I whill try to use JavaScript with C# in my UWP to use that package https://github.com/yaronn/xml-crypto in my UWP project

**Comment by JaedsonBarbosa** [comment]
So I will wait because I can't build the corefx-master with Visual Studio 2017 😞

**Comment by JaedsonBarbosa** [comment]
@karelz @krwq   I can't build the solution because I give an error, you can see the CMakeError here:
https://1drv.ms/u/s!AjDoJtNk3vvWjKIAYajeMPkWkI-m6Q
Please give me a help 🆘

**Comment by JaedsonBarbosa** [comment]
Anybody now what can I do to make it build?

**Comment by krwq** [comment]
@JaedsonBarbosa how do you build? Regular process for me is:
1. Open cmd.exe
2. `pushd corefx\repo\path`
3. `git pull`
4. `git clean -fdx` - this will clean your repository of any leftover files (be careful if you're not sure what it does)
5. `build` or `./build.sh` if you are on non-windows

**Comment by karelz** [comment]
@Jaedson33 you can also check & help us improve our [new-contributor docs](https://github.com/dotnet/corefx/wiki/New-contributor-Docs) (linked from CoreFX main page)

**Comment by JaedsonBarbosa** [comment]
I'm using the CMake 3.8.
@krwq  I do what you said, and I'm waiting about 1 hour and there is just Installing dotnet cli..., how many hours do you think I will need to wait?

**Comment by krwq** [comment]
@JaedsonBarbosa it should take up to few minutes, try restarting - it could be some connection issues, if doesn't work please file a separate issue in this repo, someone should be able to track down what has happened

**Comment by JaedsonBarbosa** [comment]
I get this error:
C:\Users\jaeds\Source\Repos\corefx>call "C:\Users\jaeds\Source\Repos\corefx\Tools\dotnetcli\dotnet.exe" restore "C:\Users\jaeds\Source\Repos\corefx\packages\microsoft.dotnet.buildtools\1.0.27-prerelease-01512-01\lib\\tool-runtime\project.csproj" --source https://dotnet.myget.org/F/dotnet-core/api/v3/index.json --source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json --source https://api.nuget.org/v3/index.json  
C:\Users\jaeds\Source\Repos\corefx\packages\microsoft.dotnet.buildtools\1.0.27-prerelease-01512-01\lib\\tool-runtime\project.csproj --source https://dotnet.myget.org/F/dotnet-core/api/v3/index.json --source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json --source https://api.nuget.org/v3/index.json
  Restoring packages for C:\Users\jaeds\Source\Repos\corefx\packages\microsoft.dotnet.buildtools\1.0.27-prerelease-01512-01\lib\tool-runtime\project.csproj...
  Retrying 'FindPackagesByIdAsync' for source 'https://dotnetmyget.blob.core.windows.net/artifacts/dotnet-buildtools/nuget/v3/flatcontainer/system.diagnostics.debug/index.json'.
  An error occurred while sending the request.
    A conexÆo com o servidor foi interrompida de modo anormal
  Retrying 'FindPackagesByIdAsync' for source 'https://dotnetmyget.blob.core.windows.net/artifacts/dotnet-buildtools/nuget/v3/flatcontainer/system.runtime/index.json'.
  An error occurred while sending the request.
    A conexÆo com o servidor foi interrompida de modo anormal
  Retrying 'FindPackagesByIdAsync' for source 'https://dotnetmyget.blob.core.windows.net/artifacts/dotnet-buildtools/nuget/v3/flatcontainer/system.text.encoding/index.json'.
  An error occurred while sending the request.
    A conexÆo com o servidor foi interrompida de modo anormal
  Retrying 'FindPackagesByIdAsync' for source 'https://dotnetmyget.blob.core.windows.net/artifacts/dotnet-buildtools/nuget/v3/flatcontainer/system.runtime/index.json'.
  An error occurred while sending the request.
    A conexÆo com o servidor foi interrompida de modo anormal
  Retrying 'FindPackagesByIdAsync' for source 'https://dotnetmyget.blob.core.windows.net/artifacts/dotnet-buildtools/nuget/v3/flatcontainer/system.text.encoding/index.json'.
  An error occurred while sending the request.
    A conexÆo com o servidor foi interrompida de modo anormal
C:\Users\jaeds\Source\Repos\corefx\Tools\dotnetcli\sdk\2.0.0-preview1-005724\NuGet.targets(97,5): error : Failed to retrieve information about 'System.Text.Encoding' from remote source 'https://dotnetmyget.blob.core.windows.net/artifacts/dotnet-buildtools/nuget/v3/flatcontainer/system.text.encoding/index.json'. [C:\Users\jaeds\Source\Repos\corefx\packages\microsoft.dotnet.buildtools\1.0.27-prerelease-01512-01\lib\tool-runtime\project.csproj]
C:\Users\jaeds\Source\Repos\corefx\Tools\dotnetcli\sdk\2.0.0-preview1-005724\NuGet.targets(97,5): error :   An error occurred while sending the request. [C:\Users\jaeds\Source\Repos\corefx\packages\microsoft.dotnet.buildtools\1.0.27-prerelease-01512-01\lib\tool-runtime\project.csproj]
C:\Users\jaeds\Source\Repos\corefx\Tools\dotnetcli\sdk\2.0.0-preview1-005724\NuGet.targets(97,5): error :   A conexÆo com o servidor foi interrompida de modo anormal [C:\Users\jaeds\Source\Repos\corefx\packages\microsoft.dotnet.buildtools\1.0.27-prerelease-01512-01\lib\tool-runtime\project.csproj]

C:\Users\jaeds\Source\Repos\corefx>set RESTORE_ERROR_LEVEL=1 
ERROR: An error occured when running: '"C:\Users\jaeds\Source\Repos\corefx\Tools\dotnetcli\dotnet.exe" restore "C:\Users\jaeds\Source\Repos\corefx\packages\microsoft.dotnet.buildtools\1.0.27-prerelease-01512-01\lib\\tool-runtime\project.csproj"'. Please check above for more details.

**Comment by karelz** [comment]
@JaedsonBarbosa please file a new issue as suggested above. Just a reminder that we may not be able to provide as much troubleshooting support to you before 5/10 as we would like to.

**Comment by JaedsonBarbosa** [comment]
It's now working 😄 , but it's toooooooooo slow, I've been waiting about 4 hours and still not over 😒 
The only think I did to make it build was move the files from C:\Users\jaeds\Source\Repos\corefx-master to C:\Users\jaeds\Source.

**Comment by JaedsonBarbosa** [comment]
Hi, now you can use this package in a UWP app, here a sample app: https://github.com/JaedsonBarbosa/corefx/tree/BigOptimization/src/System.Security.Cryptography.Xml/TesteAssinatura

**Comment by karelz** [comment]
@JaedsonBarbosa great! Is there anything from your work which would be valuable to flow back into CoreFX? (i.e. things which are not temporary hacks)

**Comment by JaedsonBarbosa** [comment]
Well, you can use my project to see what can you simplify (or remove) in the CoreFX 😄

**Comment by leopignataro** [comment]
Hello everybody, great effort on porting this!

May I ask why is the Nuget package referencing .NET Core 2.0, instead of .NET Standard 2.0? Wouldn't that be referred?

**Comment by danmosemsft** [comment]
That should have been done (c4650c9730861c61c648a6b7f1bbf40e5dfbffae)

I'm guessing you're looking at the official preview on Nuget
https://www.nuget.org/packages/System.Security.Cryptography.Xml/4.4.0-preview1-25305-02

and it's only got to Myget 
https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Security.Cryptography.Xml/4.4.0-preview2-25316-02

**Comment by leopignataro** [comment]
Yes, that's were I was looking. Thank you, @danmosemsft!

**Comment by danmosemsft** [comment]
@leopignataro no problem, I encourage you to try bits from "head" -- you can get them from the homepage here https://github.com/dotnet/cli ... you can just download the zip if you want. Let us know if you find issues -- we have only a little time left to fix them before release.

**Comment by karelz** [comment]
FYI: Here's info about timelines: https://github.com/dotnet/corefx/issues/17619#issuecomment-301937346

**Comment by leopignataro** [comment]
Since you mention it, @danmosemsft, there's this one issue:

https://github.com/dotnet/corefx/issues/19198

I've suggested a fix on the thread but my message may have been missed in all the buzz.

**Comment by karelz** [comment]
@leopignataro fix for what where? If it is fix for #19198, then it should be tracked in the bug. If it is something else, I would prefer to see separate issue for it.
If you think your fix suggestion was overlooked somewhere, raise it again on that thread and ask for feedback.

**Comment by fletchsod-developer** [comment]
I'm confused now.  I thought System.Security.Cryptography.Xml NuGet package is for .NET Framework and it is already included in Dot Net Core v2.  Did I hear this incorrectly?  Thanks.

**Comment by bartonjs** [comment]
@fletchsod-developer The package is mainly for .NET Core.  But if you reference it from a library targeting .NET Standard it will unify with the .NET Framework version on .NET Framework, and run the code from within the package on .NET Core.

We don't have any plans to put SignedXml into the default install experience for .NET Core; it's niche enough that being a separate package on NuGet seems the best distribution model.
